Align Java Executor Service behavior for `shuttingdown?`, `shutdown?`

bensheldon opened the pull request:
Fixes #1041 

- `shuttingdown?` now uses `isShutdown && !isTerminated`
- `shutdown?` now accurately represents when all tasks have terminated
- ~~`kill` now waits for all tasks to terminate (it waits indefinitely; I'm not sure if it's possible for that to hang on Java)~~ Extracted `#kill` change to #1044

eregon commented on lib/concurrent-ruby/concurrent/executor/java_executor_service.rb:
```diff
@@ -46,6 +46,7 @@ def shutdown
       def kill
         synchronize do
           @executor.shutdownNow
+          ok = @executor.awaitTermination(60, java.util.concurrent.TimeUnit::SECONDS) until ok
```
Could you reuse `wait_for_termination` here?
Also I think this line should be moved the `synchronize`, it's typically problematic to do any blocking call in `synchronize`.

bensheldon commented:
1. Something in here is causing a deadlock on JRuby in the TimerSet tests (I'm guessing it's the wait_for_termination)
2. We need a maximum Actions workflow timeout of not 6 hours.

eregon commented:
Could you make a PR to add timeouts of 10 minutes (which looks enough from looking at previous runs) for all CI workflows?

eregon commented:
Can you reproduce the issue locally with TimerSet on JRuby?
I guess it might be related to that `kill` change.
So it might be worth to extract that change to a separate PR.
In general I wonder if it's really worth the trouble & complexity of having different implementations for JRuby when there is already a pure-Ruby implementation of some concurrent-ruby class.

bensheldon commented:
@eregon Thank you for the help! 🙇🏻 

- Extracted `kill` to #1042 
- Added GitHub actions timeout in #1043 

> In general I wonder if it's really worth the trouble & complexity of having different implementations for JRuby when there is already a pure-Ruby implementation of some concurrent-ruby class.

I feel mixed. A lot of Concurrent Ruby is inspired by Java objects, so it seems like it would be _better_ (simpler? more performant? less potential for bikeshedding or 2nd-systeming them?) to lightly wrap them rather than use a Ruby implementation. I have similar vague feelings about Native C Extensions too; which to chit-chat also makes me wonder if YJIT would make that less worthwhile.

eregon commented:
> I have similar vague feelings about Native C Extensions too; which to chit-chat also makes me wonder if YJIT would make that less worthwhile.

The concurrent-ruby C ext is [only for atomic variables](https://github.com/ruby-concurrency/concurrent-ruby/blob/8b9b0da4a37585ce5eb71516aca55e93bde39115/ext/concurrent-ruby-ext/rb_concurrent.c).
Those are likely quite a bit faster than [using a Mutex](https://github.com/ruby-concurrency/concurrent-ruby/blob/8b9b0da4a37585ce5eb71516aca55e93bde39115/lib/concurrent-ruby/concurrent/atomic/mutex_atomic_boolean.rb#L8).
